### PR TITLE
CR-263 eBayes and topTable

### PR DESCRIPTION
**Pull request by EpigeneMax**
## Description
See [Jira card](https://epigenelabsteam.atlassian.net/browse/CR-263)

## Motivation and Context
Splitting the port of `limma` into smaller PR to ease the integration process

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] No code (non-breaking change that does not impact code: formatting, build system, documentation...)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] My changes are in a separate branch named after a Jira card, or -- if not applicable -- with a descriptive name.
- [x] I have descriptive commit messages, with a short title (first line).
- [x] I have added and/or updated tests to reflect my changes.
- [x] I have reviewed and updated the documentation accordingly.
- [x] I have updated the [Changelog](../CHANGELOG.md).
- [x] My code and documentation follow the [numpy docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html).
- [x] I have reviewed possible licensing issues, preferably with a person in charge of the topic.

### CODE REVIEW

File: inmoose/limma/decidetests.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+            labels = levels
+        tab.index = labels
+        tab.columns = self.columns
+        for i in range(nlevels):
+            tab.iloc[i, :] = np.nan_sum(self == levels[i], axis=1)
+        return tab
+
+
+class FStat(np.ndarray):
+
```
Comment (Lea-Meunier) [inline inmoose/limma/decidetests.py:80]: maybe add some docstring here

File: inmoose/limma/decidetests.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    def _constructor_sliced(self):
+        return pd.Series
+
+    def __init__(self, df, *args, **kwargs):
+        super().__init__(df, *args, **kwargs)
+        self.levels = None
+        self.labels = None
+
+    def summary(self):
+        """ """
```
Comment (Lea-Meunier) [inline inmoose/limma/decidetests.py:63]: the docstring here is empty

File: inmoose/limma/lmfit.py [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+        # )
+    else:
+        if ndups < 2 and block is None:
+            fit = lm_series(
+                y.exprs, design=design, ndups=ndups, spacing=spacing, weights=weights
+            )
+        else:
+            if correlation is None:
+                raise ValueError(
+                    "the correlation must be set, see function duplicateCorrelation"
```
Comment (Lea-Meunier) [inline inmoose/limma/lmfit.py:232]: Just to be sure, `duplicateCorrelation` function is not yet implemented and will may be in a future PR? Or it come from another module? (I couldn't find it in the code) 
Maybe we can put an `NotImplementedError`

**Review by Lea-Meunier — COMMENTED** [review]
Impressive work! The PR is very complex I'm not sure my review is very effective, it looks fine to me, I just noticed 2 docstrings missing

File: inmoose/limma/lmfit.py (hunk 3 above)
Comment (EpigeneMax) [reply inmoose/limma/lmfit.py]: This `ValueError` informs the user that the `correlation` parameter is mandatory for duplicated channels.
However, the duplicated channels features are not implemented in `inmoose` for the time being.
The following line of code does raise a `NotImplementedError` to signal this.

File: inmoose/limma/decidetests.py (hunk 1 above)
Comment (EpigeneMax) [reply inmoose/limma/decidetests.py:99]: fixed

File: inmoose/limma/decidetests.py (hunk 2 above)
Comment (EpigeneMax) [reply inmoose/limma/decidetests.py]: fixed

File: inmoose/limma/contrasts.py [hunk 4]
```diff
@@ -243,16 +243,17 @@ def contrasts_fit(fit, contrasts=None, coefficients=None):
...
         ngenes = fit.stdev_unscaled.shape[0]
         ncont = contrasts.shape[1]
         U = np.ones((ngenes, ncont))
+        U = pd.DataFrame(U, index=fit.stdev_unscaled.index, columns=contrasts.columns)
         o = np.ones((1, ncoef))
         for i in range(ngenes):
-            RUC = R @ (fit.stdev_unscaled[i, :][:, None] * contrasts)
-            U[i, :] = np.sqrt(o @ RUC**2)
+            RUC = R @ (fit.stdev_unscaled.values[i, :][:, None] * contrasts)
+            U.iloc[i, :] = np.sqrt(o @ RUC**2)
```
Comment (guillaumeap) [inline inmoose/limma/contrasts.py:256]: Not that important but I think numpy is faster than pandas, maybe you can create the DataFrame object after the for loop?

File: inmoose/limma/decidetests.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        return obj
+
+    def __array_finalize__(self, obj):
+        if obj is None:
+            return
+        self.df1 = getattr(obj, "df1", None)
+        self.df2 = getattr(obj, "df2", None)
+
+
+def classifyTestsF(self, cor_matrix=None, df=np.inf, p_value=0.01, fstat_only=False):
```
Comment (guillaumeap) [inline inmoose/limma/decidetests.py:113]: Is it ok to call a function argument self?

File: inmoose/limma/lmfit.py (hunk 3 above)
Comment (Lea-Meunier) [reply inmoose/limma/lmfit.py]: very clear thanks !

File: inmoose/limma/ebayes.py [hunk 6]
```diff
@@ -19,9 +19,282 @@
...
+    return fit
+
+
+def _ebayes(
+    fit,
+    proportion=0.01,
+    stdev_coef_lim=(0.1, 4),
+    trend=False,
+    robust=False,
+    winsor_tail_p=(0.05, 0.1),
```
Comment (guillaumeap) [inline inmoose/limma/ebayes.py:217]: the type of winsor_tail_p changes between the two functions

**Review by guillaumeap — COMMENTED** [review]
Very impressive PR! Like Léa, I'm not sure that I can check the statistics but the first tests I made look pretty good!
I just have a few comments

File: inmoose/limma/decidetests.py (hunk 5 above)
Comment (EpigeneMax) [reply inmoose/limma/decidetests.py:113]: `self` is just a name, it has no special meaning.
When calling a method on an object `obj.method(x,y,z)`, under the hood that call is transformed to `method(obj, x, y, z)`, hence the convention is to call `self` the first argument of methods.

Now you can also define methods outside of the class body, _e.g._ to break down a big class into smaller chunks, following this pattern:
```python
def f(self, x, y):
    return x + y + self.z

class MyClass:
    import f
```
In that case, calling `self` the first argument of `f` is a way to remind the reader that `f` is intended to be a method, even though it is defined out-of-scope.

File: inmoose/limma/ebayes.py (hunk 6 above)
Comment (EpigeneMax) [reply inmoose/limma/ebayes.py:218]: hawk eye
fixed

**Review by guillaumeap — APPROVED** [review]
Congrats!

**Pull request merged**